snuba/lw_deletions/strategy.py

```diff
@@ -0,0 +1,158 @@
+import typing
+from typing import Mapping, Optional, Sequence, TypeVar
+
+import rapidjson
+from arroyo.backends.kafka import KafkaPayload
+from arroyo.processing.strategies import (
+    CommitOffsets,
+    ProcessingStrategy,
+    ProcessingStrategyFactory,
+)
+from arroyo.processing.strategies.abstract import MessageRejected
+from arroyo.types import BaseValue, Commit, Message, Partition
+
+from snuba import settings
+from snuba.attribution import AppID
+from snuba.attribution.attribution_info import AttributionInfo
+from snuba.datasets.storage import WritableTableStorage
+from snuba.lw_deletions.batching import BatchStepCustom, ValuesBatch
+from snuba.lw_deletions.formatters import Formatter
+from snuba.query.query_settings import HTTPQuerySettings
+from snuba.state import get_int_config
+from snuba.web.bulk_delete_query import construct_or_conditions, construct_query
+from snuba.web.delete_query import (
+    ConditionsType,
+    TooManyOngoingMutationsError,
+    _execute_query,
+    _num_ongoing_mutations,
+)
+
+TPayload = TypeVar("TPayload")
+
+import logging
+
+logger = logging.Logger(__name__)
+
+
+class FormatQuery(ProcessingStrategy[ValuesBatch[KafkaPayload]]):
+    def __init__(
+        self,
+        next_step: ProcessingStrategy[ValuesBatch[KafkaPayload]],
+        storage: WritableTableStorage,
+        formatter: Formatter,
+    ) -> None:
+        self.__next_step = next_step
+        self.__storage = storage
+        self.__cluster_name = self.__storage.get_cluster().get_clickhouse_cluster_name()
+        self.__tables = storage.get_deletion_settings().tables
+        self.__formatter: Formatter = formatter
+
+    def poll(self) -> None:
+        self.__next_step.poll()
+
+    def submit(self, message: Message[ValuesBatch[KafkaPayload]]) -> None:
+        decode_messages = [
+            rapidjson.loads(m.payload.value) for m in message.value.payload
+        ]
+        conditions = self.__formatter.format(decode_messages)
+
+        try:
+            self._execute_delete(conditions)
+        except TooManyOngoingMutationsError:
+            raise MessageRejected
```

Comment:
What is the intended behavior when too many mutations are ongoing and a message is rejected? Probably better to clarify this behavior in a comment.

Reply:
@onkar raising `MessageRejected` is the way in which we pause the consumer when we want to apply back pressure. Comes from https://github.com/getsentry/arroyo/blob/a8545c79bf81d78ff432addf9c8ad737b81082a7/arroyo/processing/processor.py#L417

But I can add a comment clarifying that this will cause back pressure